openedx_events/event_bus/__init__.py

```diff
@@ -1 +1,83 @@
 """Classes and utility functions for the event bus."""
+
+from abc import ABC, abstractmethod
+from typing import NoReturn, Optional
+
+from django.conf import settings
+from django.utils.module_loading import import_string
+
+
+class EventBus(ABC):
+    """
+    Parent class for event bus implementations.
+    """
+
+    @abstractmethod
+    def send(
+            *, signal: OpenEdxPublicSignal, topic: str, event_key_field: str, event_data: dict,
+            sync: bool = False,
+    ) -> None:
+        """
+        Send a signal event to the event bus under the specified topic.
+
+        Arguments:
+            signal: The original OpenEdxPublicSignal the event was sent to
+            topic: The event bus topic for the event
+            event_key_field: Path to the event data field to use as the event key (period-delimited
+              string naming the dictionary keys to descend)
+            event_data: The event data (kwargs) sent to the signal
+            sync: Whether to wait indefinitely for event to be received by the message bus (probably
+              only want to use this for testing)
+        """
+        pass
+
+    @abstractmethod
+    def consume_indefinitely(*, topic: str, group_id: str) -> NoReturn:
+        """
+        Consume events from a topic in an infinite loop.
+
+        Events will be converted into calls to Django signals.
+
+        Arguments:
+            topic: The event bus topic to consume from.
+            group_id: The consumer group to participate in.
+        """
+        pass
+
+
+class NoEventBus(EventBus):
+    """
+    Stub implementation to "load" when no implementation is properly configured.
+    """
+
+    def send(
+            *, signal: OpenEdxPublicSignal, topic: str, event_key_field: str, event_data: dict,
+            sync: bool = False,
+    ) -> None:
+        pass
+
+    def consume_indefinitely(*, topic: str, group_id: str) -> NoReturn:
+        pass
+
+
+def _load_event_bus() -> EventBus:
+    """
+    Load the currently configured event bus implementation.
+
+    If configuration is invalid, will return a NoEventBus instance that does nothing.
+    """
+    try:
+        eb = settings.EVENT_BUS
+        loader = import_string(eb['loader'])
+        impl = loader(eb)
```

Comment:
[author note] I'm imagining that the `EVENT_BUS` dictionary setting might carry the additional implementation-specific settings that the loader function will then want to read. So we might have:

```
EVENT_BUS = {
  "loader": "edx_event_bus_kafka.create",
  "env_prefix": "prod",
  "kafka": {
    "schema_registry": "...",
    "bootstrap_servers": "..."
  }
}
```

Reply:
Generally I think this is ok, but I find the naming a bit confusing.
1. Instead of "loader" setting, could it just be "event_bus_class" or something? What is `create`, and could we just use provide a path to the class instead?
2. Do we need to pass the settings to the "loader"? Could we use something like the DRF settings, as noted in the other comment?

Reply:
1. We could just use the class constructor, yeah. So `edx_event_bus_kafka.EventBusKafka`.
2. I suppose we don't need to pass the settings. The implementation could look them up in `django.conf.settings` itself.

Reply:
In 48e3918 it is now `EVENT_BUS_PRODUCER` and `EVENT_BUS_CONSUMER` rather than `loader`. No more passing settings. Loader can be either a class (acting as a constructor) or a function.